[마켓불리 김원희] Add: 장바구니 Endpoint 구현 완료 (리뷰 반영하여 수정, 컨플릭트 수정, 3차 수정)

supersaloon opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- 장바구니 CRUD Endpoint 구현 완료
		 

## 수정 사항들 자세한 내용
- 장바구니 상품 추가
  - 장바구니에 사용자가 선택한 상품이 이미 존재하는 경우, 장바구니의 해당 상품 수량을 올림
- 장바구니 상품 조회
- 장바구니 상품 삭제
- 장바구니 상품 수량 수정
  - 프론트에서 "plus"를 전달받으면 장바구니 상품의 수량 += 1
  - 프론트에서 "minus"를 전달받으면 장바구니 상품의 수량 -= 1


## 기타 질문 및 특이 사항
- 장바구니 상품 추가 관련하여서 아래와 같은 모델을 이용하여 다음과 같은 로직을 사용해 보았는데 적절한 로직인 것인지 궁금합니다.

![장바구니 질문](https://user-images.githubusercontent.com/69733417/102733598-06de3480-4381-11eb-9db5-f3c316de1aba.png)

  - 모델 설명
    - order에서 product로 cart라는 중간테이블을 통해서 Many to Many로 연결
    - cart 테이블의 각 인스턴스는 order 인스턴스를 바라봄

  - 로직 설명
    1. 장바구니에 상품 추가 요청을 받으면 order 테이블에 user_id가 사용자의 id이고, status=1(장바구니)인 instance가 있는지 확인
    2. 없으면 생성하여 가져오고, 있으면 기존의 instance를 사용
    3. 위에서 가져온 order(user_id=user_id, status_id=1(장바구니))의 instance_id를 바라보는 instance를 cart(중간) 테이블에 생성(instance 하나가 하나의 상품 항목)
 
 이상과 같이 로직을 짜 보았는데, 자연스러운 로직인지 그리고 혹시 더 좋은 방법이 있다면 가이드 부탁드립니다.
 
 감사합니다 :) 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
Add: 장바구니 기능 구현
- 장바구니에 상품 추가 구현
  - 장바구니에 추가하려는 상품이 이미 존재하는 경우 기존 상품의 수량을 증가
- 장바구니 목록 조회 구현
- 장바구니 수량 수정 구현
- 장바구니 상품 삭제 구현
- 아직 데코레이터가 구현되지 않은 상태이여서 사용자는 임의로 지정함

soheon-lee commented:
원희님 장바구니 모델링 완전 좋습니다~

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+from django.http      import JsonResponse
+from django.views     import View
+
+from order.models   import Cart, Order
+from user.models      import Address
```
`import` 기준으로 정렬해주세요 :)

supersaloon replied:
`import` 기준으로 정렬 수정하였습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+            # 사용자의 장바구니에 이미 해당 상품이 있는 경우 전달된 quantity 만큼 그 수량을 올림
+            if Cart.objects.filter(order_id=order_cart_id, product_id=data['product_id']).exists():
+                product_in_cart = Cart.objects.get(order_id=order_cart_id, product_id=data['product_id'])
+                product_in_cart.quantity += int(data['quantity'])
+                product_in_cart.save()
```
수량 변경 로직 good!

supersaloon replied:
감사합니다 :)

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+            user_id = "3"  # 데코레이터 나오기 전까지 사용자 임의지정
+
+            # 사용자 장바구니 가져오기, 있으면 가져오고 없으면 생성하여 가져옴
+            if not Order.objects.filter(user_id=user_id, status_id=1).exists():
+                order_cart = Order.objects.create(
+                    user_id           = user_id,
```
```suggestion
            if not Order.objects.filter(user_id=user_id, status_id=1).exists():
                new_order = Order.objects.create(
                    user_id           = user_id,
```
Cart는 이 아래 32번째 라인부터 다루기 시작하는 것으로 판단되고, 19번 라인은 주문만을 다루는 것 같은데 변수명에 cart가 포함되어 있으면 다소 헷갈릴 위험이 있을 것 같습니다 :) 따라서 보다 명확하게 주문이라는 것을 알려줄 수 있는 변수명을 추천드립니다. 꼭 `new_order`로 하실 필요는 없고, 의미에 가장 적합하다고 생각되는 단어로 사용해주세요

supersaloon replied:
order_cart라는 변수명은 order의 instance인데, status=1(장바구니)인 instance(이 인스턴스를 cart의 instance들이 바라봄)라는 의미에서 사용해 보았습니다.

여기서는 new_order로 정정하도록 하겠습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+
+    # @login_required
+    def delete(self, request):
+        try:
+            data = json.loads(request.body)
+            item = Cart.objects.get(id=data['cart_item_id'])
+            item.delete()
+
+            return JsonResponse({'MESSAGE': 'SUCCESS'}, status=204)
+
+        except KeyError as e:
+            return JsonResponse({"MESSAGE": "KEY_ERROR => " + e.args[0]}, status=400)
+        except Cart.DoesNotExist:
+            return JsonResponse({"MESSAGE": "ITEM_DOES_NOT_EXIST"}, status=400)
```
한가지 질문이 있는데, 혹시 카트에 담긴 마지막 상품까지 지워진다면, 연결된 주문도 사라지나요? 아니면 주문은 계속 살아있나요?

supersaloon replied:
주문은 계속 살리려고 의도했습니다.

제가 생각한 로직으로는 항상 status=1(장바구니)인 order의 instance가 존재하고,
그 order의 instance(user_id=사용자 id, status=1(장바구니))를 cart(product와 order의 중간테이블)의 instance들이 바라보는 상태가 장바구니라고 판단했습니다.

하여, 회원가입할 때 강제적으로 장바구니 instance(user_id=사용자 id, status=1(장바구니))를 생성할까도 고민해보았으나, 우선은 order 앱에서 장바구니에 상품을 추가할 때 검사하고 없으면 생성하는 방식으로 구현해 보았습니다.

soheon-lee replied:
원래 의도하신 바라면 좋습니다!

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+                cart_item.quantity -= 1
+            elif delta == "plus":
+                cart_item.quantity += 1
+            else:
+                return JsonResponse({"MESSAGE": "KEY_ERROR"}, status=400)
```
이 키에러는 `delta`라는 키 자체가 안들어왔을 때 나는 에러여야 하고, `delta`라는 키는 들어왔는데 그 값이 `minus`나 `plus`가 아닌 경우라면 에러 메세지가 달라지는 것이 좋을 것 같습니다 :)

supersaloon replied:
delta라는 키는 받았으나, 그 값이 `minus`나 `plus`가 아니라는 의미로 수정하였습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+            user_id = "3"  # 데코레이터 나오기 전까지 사용자 임의지정
+
+            # 사용자 장바구니 가져오기, 있으면 가져오고 없으면 생성하여 가져옴
+            if not Order.objects.filter(user_id=user_id, status_id=1).exists():
+                order_cart = Order.objects.create(
+                    user_id           = user_id,
+                    order_number      = user_id + "-" + datetime.now().strftime('%Y-%m-%d-%H-%M-%S-%f'),
+                    status_id         = 1,
+                    address_id        = Address.objects.get(user_id=user_id, is_active=1).id,
+                    payment_method_id = 1
+                )
+                order_cart_id = order_cart.id
+            else:
+                order_cart_id = Order.objects.get(user_id=user_id, status_id=1).id
+
+            # 사용자의 장바구니에 이미 해당 상품이 있는 경우 전달된 quantity 만큼 그 수량을 올림
+            if Cart.objects.filter(order_id=order_cart_id, product_id=data['product_id']).exists():
+                product_in_cart = Cart.objects.get(order_id=order_cart_id, product_id=data['product_id'])
+                product_in_cart.quantity += int(data['quantity'])
```
Order 테이블과 Cart테이블을 동시에 다루어야 하는 로직이라, 헷갈릴 수도 있었을 텐데, 차분히 순서대로 정말 잘 작성해주셨네요 원희님!

이제는 transaction을 도입해야할 때가 왔습니다. 작업 도중 중간에 에러가 나거나, 알 수 없는 서버 결함으로 인하여 Order만 생성되고 그에 따른 Cart가 생성되지 않은 채 기능이 끝났을 때, 애매하게 Order만 생성되는 것이 아니라, Order 데이터와 Cart데이터가 깔끔하게 둘 다 생성되지 않도록 transaction을 적용해주세요. 아마 이해하셨을거라 생각됩니다.

supersaloon replied:
원자성을 보장할 수 있도록 post 메소드 자체에 데코레이터를 이용한 트랜젝션을 적용하였습니다.

`with`를 사용하는 방식도 고민해 보았으나,
`with`를 로 특정 구간을 나누는 것이 post 메소드(order 처리 -> cart 처리 -> 끝)에는 의미가 없어보여 데코레이터로 post 메소드 전체에 트랜젝션을 적용하였습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+    # @login_required
+    def patch(self, request):
+        try:
+            data = json.loads(request.body)
+            delta = data['delta']
+            cart_item = Cart.objects.get(id=data['cart_item_id'])
```
코드 간격 정렬해주세요 :)

supersaloon replied:
간격에 맞게 정렬하였습니다.

Commit:
Modify: PR #7에 대한 리뷰 반영하여 수정
- 컨벤션 맞지 않는 부분 수정
- post 메소드에 트랜젝션 적용
- patch 메소드의 키 delta를 받는 부분에서 에러메세지가 모호하던 부분 명확히 수정

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,112 @@
+        try:
+            data = json.loads(request.body)
+            user_id = "3"  # 데코레이터 나오기 전까지 사용자 임의지정
+
+            # 사용자 장바구니 가져오기, 있으면 가져오고 없으면 생성하여 가져옴
+            if not Order.objects.filter(user_id=user_id, status_id=1).exists():
+                new_order = Order.objects.create(
+                    user_id           = user_id,
+                    order_number      = user_id + "-" + datetime.now().strftime('%Y-%m-%d-%H-%M-%S-%f'),
+                    status_id         = 1,
+                    address_id        = Address.objects.get(user_id=user_id, is_active=1).id,
+                    payment_method_id = 1
+                )
+                new_order_id = new_order.id
+            else:
+                new_order_id = Order.objects.get(user_id=user_id, status_id=1).id
+
+            # 사용자의 장바구니에 이미 해당 상품이 있는 경우 전달된 quantity 만큼 그 수량을 올림
```
곳곳에 메모용 주석이 있습니다 원희님. 로직을 간결히 설명하시는 용도라면 좋습니다 👍 
그러나 merge 전에는 해당 주석은 지워주시는 것이 좋을 것 같습니다.

코드에서의 주석은 코드 자체만으로 이해가 어려울 때 이해를 돕는 용도인데, 현재 로직은 너무 어려운 기능이 아니라고 판단되고, 또 따로 설명이 필요하지 않을만큼 코드 이해하기 쉽게 잘 짜주셨기 때문에 제거해주셔도 좋을 것 같네요 :) 파일에 있는 다른 주석들도 함께 정리해주세요

supersaloon replied:
아, 넵 주석 지우도록 하겠습니다.

soheon-lee commented on order/views.py:
```diff
@@ -1,3 +1,112 @@
+                product_in_cart.quantity += int(data['quantity'])
+                product_in_cart.save()
+                return JsonResponse({'MESSAGE': 'SUCCESS'}, status=200)
+            else:
+                Cart.objects.create(
+                    order_id   = new_order_id,
+                    product_id = data['product_id'],
+                    quantity   = data['quantity'],
+                )
+                return JsonResponse({'MESSAGE': 'SUCCESS'}, status=201)
```
`else`가 꼭 필요한 구문이 아닌 것 같습니다 :)

supersaloon replied:
`else` 절 삭제하였습니다.

Commit:
Modify: #7 리뷰 반영하여 수정(2차)
- 주석 제거
- 불필요한 else 구문 삭제

Commit:
Modify: conflict 수정
- urls.py
- 0003_detailedimage_productdescription.py

Commit:
conflict 수정 2차

jamessoun93 commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+            data = json.loads(request.body)
+            user_id = "3"
+
+            if not Order.objects.filter(user_id=user_id, status_id=1).exists():
+                new_order = Order.objects.create(
+                    user_id           = user_id,
+                    order_number      = user_id + "-" + datetime.now().strftime('%Y-%m-%d-%H-%M-%S-%f'),
+                    status_id         = 1,
+                    address_id        = Address.objects.get(user_id=user_id, is_active=1).id,
+                    payment_method_id = 1
+                )
+                new_order_id = new_order.id
+            else:
+                new_order_id = Order.objects.get(user_id=user_id, status_id=1).id
```
장고의 `get_or_create()` 메소드를 통해 불필요한 `if ... else ...` 문을 줄일 수 있어보이네요! 한번 적용해보세요~

supersaloon replied:
기존의 .filter로 instance 조회하여 유무에 따라 if else문으로 분기하는 대신, get_or_create() 적용하여서,
Order 존재 유무에 대한 flag를 받고 flag가 True, False인 경우에 대해 각각 분기하는 방식으로 코드 수정하였습니다.

jamessoun93 replied:
확인했습니다 👍

jamessoun93 commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+
+class CartView(View):
+    # @login_required
+    @transaction.atomic
```
소헌님 리뷰반영하여 transaction 적용하셨군요 👍 👍 👍

supersaloon replied:
감사합니다 : )

jamessoun93 commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+    def post(self, request):
+        try:
+            data = json.loads(request.body)
+            user_id = "3"
```
이 부분은 수정내역을 확인하니 아직 추가 구현중인 부분이라 임시로 정해두신걸로 확인되는데, 언제쯤 완료가 될까요?

supersaloon replied:
사용자 인증 관련 데코레이터는 현주님께서 개발 중이신데, 구현 및 테스트는 어제 완료되었습니다. 오늘 중으로 Pull Request 하는 것으로 알고 있습니다.

jamessoun93 replied:
확인했습니다 👍

jamessoun93 commented on order/views.py:
```diff
@@ -1,3 +1,110 @@
+    # @login_required
+    def get(self, request):
+        try:
+            user_id = 3
+            order_cart_id = Order.objects.get(user_id=user_id, status=1).id
+            cart = Cart.objects.filter(order_id=order_cart_id).prefetch_related("product__discount", "product__packing_type")
```
`user_id`가 프론트에서 넘어오는 상황인경우 불필요하게 `Order` 테이블에서 시작하여 해당 `user_id` 를 가지고 있는 `Order` 객체를 찾아온 후, 해당 `Order` 객체의 `id` 를 찾아 또 `Cart` 테이블에서 필터를 하기보다는, 

처음부터 원하는 데이터가 `Cart` 테이블에 있으니, `Cart` 테이블에서 시작하여 장고의 `field lookup`을 활용하여 아래처럼 필터링 해줄 수 있습니다.
```python
Cart.objects.filter(order__user_id=user_id, order__status=1).prefetch_related(...)
```

supersaloon replied:
기존 `Order` 테이블에서 시작하여 `user_id`를 이용하여 `Cart` 테이블로 이동 후 `.filter`하는 방식에서,
시작점을 `Cart` 테이블로 하고 Django의 `field lookup` 활용하는 방식으로 코드 수정하였습니다.

이 부분을 좀 더 효율적으로 하려면 어떤방식으로 해야할지에 대해서 고민을 많이 했던 로직인데, 디테일한 가이드 감사드립니다 : )

jamessoun93 replied:
확인했습니다 👍

Commit:
Modify: #7 Pull Request 에 대한 리뷰 반영하여 수정 3차
- 사용자의 장바구니 Order instance의 유무를 get_or_create 활용하여 검사하도록 수정
- Cart 테이블의 데이터를 받아오기 위해서
  - 기존 Order 테이블을 진입점으로 쿼리하였으나,
  - Cart 테이블을 진입점으로 변경하고
  - field lookup 으로 데이터 조회하도록 수정

Commit:
Modify: conflict 수정
- market_bully/urls.py

supersaloon committed:
Merge branch 'master' into feature/order-cart

jamessoun93 commented on order/views.py:
```diff
@@ -1,3 +1,109 @@
+
+            new_order_id = new_order.id
+
+            if Cart.objects.filter(order_id=new_order_id, product_id=data['product_id']).exists():
+                product_in_cart = Cart.objects.get(order_id=new_order_id, product_id=data['product_id'])
+                product_in_cart.quantity += int(data['quantity'])
+                product_in_cart.save()
+                return JsonResponse({'MESSAGE': 'SUCCESS'}, status=200)
+
+            Cart.objects.create(
+                order_id   = new_order_id,
+                product_id = data['product_id'],
+                quantity   = data['quantity'],
+            )
```
여기는 `update_or_create()` 적용해보세요~

jamessoun93 commented:
LGTM 이 머지 않았습니다! 장고의 여러가지 메소드를 활용해보게끔 해보고 싶은 욕심이 있어서 추가 코멘트 남겨드렸습니다 👍

jamessoun93 commented:
LGTM 👍
